src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,199 @@
+                }
+
+                if (!allCompilationAssemblies.Any(usedAssemblyLookup.Contains))
+                {
+                    // None of the assemblies brought into this compilation are in the remaining
+                    // used assemblies list, so we will consider the reference unused.
+                    unusedReferencesBuilder.Add(reference);
+                    continue;
+                }
+
+                // Remove all assemblies that are brought into this compilation by this reference.
+                usedAssemblyLookup.ExceptWith(allCompilationAssemblies);
+            }
+
+            return;
+
+            static ImmutableArray<string> GetAllCompilationAssemblies(ReferenceInfo reference)
+            {
+                var transitiveCompilationAssemblies = reference.Dependencies
+                    .SelectMany(dependency => GetAllCompilationAssemblies(dependency));
+                return reference.CompilationAssemblies
+                    .Concat(transitiveCompilationAssemblies)
+                    .ToImmutableArray();
+            }
+        }
+
+        public async Task<Project> UpdateReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken)
+        {
+            var referenceCleanupService = project.Solution.Workspace.Services.GetRequiredService<IReferenceCleanupService>();
+
+            await ApplyReferenceUpdatesAsync(referenceCleanupService, project.FilePath!, referenceUpdates, cancellationToken).ConfigureAwait(false);
+
+            return project.Solution.Workspace.CurrentSolution.GetRequiredProject(project.Id);
+        }
+
+        internal static async Task ApplyReferenceUpdatesAsync(
+            IReferenceCleanupService referenceCleanupService,
+            string projectFilePath,
+            ImmutableArray<ReferenceUpdate> referenceUpdates,
+            CancellationToken cancellationToken)
+        {
+            foreach (var referenceUpdate in referenceUpdates)
+            {
+                // If the update action would not change the reference, then
+                // continue to the next update.
+                if (referenceUpdate.Action == UpdateAction.TreatAsUnused &&
+                    !referenceUpdate.ReferenceInfo.TreatAsUsed)
+                {
+                    continue;
+                }
+                else if (referenceUpdate.Action == UpdateAction.TreatAsUsed &&
+                    referenceUpdate.ReferenceInfo.TreatAsUsed)
+                {
+                    continue;
+                }
+                else if (referenceUpdate.Action == UpdateAction.None)
+                {
+                    continue;
+                }
+
+                await referenceCleanupService.TryUpdateReferenceAsync(
+                    projectFilePath,
+                    referenceUpdate,
+                    cancellationToken).ConfigureAwait(false);
```

Comment:
Hmm, does the ConfigureAwait(false) here interfere with the JTF.Run() being used at the VS layer?